image/local.go

```diff
@@ -270,16 +282,21 @@ func (l *local) Save() (string, error) {
 	repoName := t.String()
 
 	pr, pw := io.Pipe()
+	defer pw.Close()
 	go func() {
 		res, err := l.Docker.ImageLoad(ctx, pr, true)
-		if err == nil {
-			io.Copy(ioutil.Discard, res.Body)
-			res.Body.Close()
+		if err != nil {
+			done <- err
+			return
 		}
-		done <- err
+		defer res.Body.Close()
+		io.Copy(ioutil.Discard, res.Body)
+
+		done <- nil
 	}()
 
 	tw := tar.NewWriter(pw)
+	defer tw.Close()
```

Comment:
This is also closed further down (and there appear to be a number of double-closes in this method).

Reply:
We need to close the pipe later on in the function to terminate the go routine that is writing the tar. We wanted to defer here for the cases in which we return early due to error. As far as I know, Close is idempotent.

Reply:
This is [not always true](https://golang.org/pkg/io/#Closer), but seems okay in the case of `tar.Writer`. Okay with this for now, but I think we should consider refactoring some of this later.